NPM package updates; `downshift`, `focus-trap-react`, `immer`, `redux`, `uuid`

lb- opened the pull request:
This PR updates a set of core admin UI related packages, many of which have not been updated for a few years and are no longer maintained in their current version.

This also reduces our npm audit warnings 11 down to to 7 high severity vulnerabilities.

## Details

- Core updates to several admin user (UI) facing libraries, many of these packages are years out of date and needed some attention.
- Longer term, I think we could simplify our package landscape (especially the immer/Redux tools), for now though this PR does the minimum possible changes for us to adopt the latest versions of these packages.

### Updates applied

| Package                                   | Old Version | New Version |
| ----------------------------------------- | ----------- | ----------- |
| **Major**                                 |             |             |
| ◯ downshift                               | ^7.6.2      | ^9.0.8      |
| ◯ focus-trap-react                        | ^8.11.3     | ^10.3.1     |
| ◯ immer                                   | ^9.0.21     | ^10.1.1     |
| ◯ react-redux (v9 blocked)                | ^7.2.2      | ^8.1.3      |
| ◯ redux                                   | ^4.2.1      | ^5.0.1      |
| ◯ redux-thunk                             | ^2.4.2      | ^3.1.0      |
| ◯ reselect                                | ^4.1.8      | ^5.1.1      |
| ◯ uuid                                    | ^9.0.1      | ^11.0.5     |


### Other blocked updates
  - We are starting to be critically blocked by our Draftail dependencies relying on specific React versions, beyond that, I have been able to update a lot of libraries.
  - I did not adjust the `draft-js` package as it was causing an error when trying to update, due to  our `draftail` 2.0.0-rc usage, this will require more time and input from @thibaudcolas probably. See also https://github.com/springload/draftail/issues/213#issuecomment-826158209
  - Stylelint - waiting on PR https://github.com/wagtail/stylelint-config-wagtail/pull/45
  - Eslint - waiting on issue to be worked on https://github.com/wagtail/eslint-config-wagtail/issues/22
  - Storybook - see https://github.com/wagtail/wagtail/issues/12553

## Testing

- Comments will be impacted, smoke testing required for a set of comments behaviours, uses Redux packages, Immer and Focus trap react. Note that the bug https://github.com/wagtail/wagtail/issues/11021 is not related to this package update and is an existing bug.
- Page explorer - It uses Redux and the focus trap component.
- StreamField / Rich text - Combo box (block chooser) - This uses the Downshift library.

## Relevant upgrade release notes

- **`focus-trap-react` - v8 to v10**
  - https://github.com/focus-trap/focus-trap-react/releases/tag/v9.0.0
  - https://github.com/focus-trap/focus-trap-react/releases/tag/v10.0.0
- **`downshift` - v7 to v9**
  - https://github.com/downshift-js/downshift/releases/tag/v9.0.0
  - https://github.com/downshift-js/downshift/releases/tag/v8.0.0
  - https://github.com/downshift-js/downshift/blob/master/src/hooks/MIGRATION_V8.md
- **`uuid` - v9 to v11**
  - https://github.com/uuidjs/uuid/blob/main/CHANGELOG.md#1100-2024-10-26
  - https://github.com/uuidjs/uuid/blob/main/CHANGELOG.md#1000-2024-06-07
- **`immer` v9 to v10**
  - https://github.com/immerjs/immer/releases/tag/v10.0.0
- **Update Redux ecosystem**
  - https://github.com/reduxjs/redux-thunk/releases/tag/v3.1.0
  - https://github.com/reduxjs/redux/releases/tag/v5.0.0
  - https://github.com/reduxjs/reselect/releases/tag/v5.0.1
  - https://github.com/reduxjs/react-redux/releases/tag/v8.0.0 (note: react-redux cannot be updated further due to React dependency, which is blocked by Draftail)

---

<details>
<summary><strong>Outdated package comparison (before/after)</strong></summary>

**Note:** These are not the updates I have done (see above for that), instead this is a reference of what updates still need to be done before & after this PR.

### Before

* 31 outdated packages

| Package                            | Old Version | New Version |
| ---------------------------------- | ----------- | ----------- |
| **Major**                          |             |             |
| ◯ @storybook/addon-controls        | ^7.6.14     | ^8.4.2      |
| ◯ @storybook/addon-docs            | ^7.6.14     | ^8.4.2      |
| ◯ @storybook/react                 | ^7.6.14     | ^8.4.2      |
| ◯ @storybook/react-webpack5        | ^7.6.14     | ^8.4.2      |
| ◯ @types/react                     | ^16.14.21   | ^18.3.12    |
| ◯ @types/react-dom                 | ^16.0       | ^18.3       |
| ◯ @typescript-eslint/eslint-plugin | ^6.21.0     | ^8.13.0     |
| ◯ @typescript-eslint/parser        | ^6.21.0     | ^8.13.0     |
| ◯ css-loader                       | ^6.10.0     | ^7.1.2      |
| ◯ cssnano                          | ^5.1.15     | ^7.0.6      |
| ◯ downshift                        | ^7.6.2      | ^9.0.8      |
| ◯ eslint                           | ^8.56.0     | ^9.14.0     |
| ◯ focus-trap-react                 | ^8.11.3     | ^10.3.0     |
| ◯ immer                            | ^9.0.21     | ^10.1.1     |
| ◯ react                            | ^16.14.0    | ^18.3.1     |
| ◯ react-dom                        | ^16.14.0    | ^18.3.1     |
| ◯ react-redux                      | ^7.2.2      | ^9.1.2      |
| ◯ react-test-renderer              | ^16.14.0    | ^18.3.1     |
| ◯ react-transition-group           | ^1.1.3      | ^4.4.5      |
| ◯ redux                            | ^4.2.1      | ^5.0.1      |
| ◯ redux-thunk                      | ^2.4.2      | ^3.1.0      |
| ◯ reselect                         | ^4.1.8      | ^5.1.1      |
| ◯ sass-loader                      | ^14.1.0     | ^16.0.3     |
| ◯ storybook                        | ^7.6.14     | ^8.4.2      |
| ◯ storybook-django                 | ^0.5.1      | ^1.0.0      |
| ◯ stylelint                        | ^15.11.0    | ^16.10.0    |
| ◯ uuid                             | ^9.0.1      | ^11.0.2     |
| **Major version zero**             |             |             |
| ◯ @types/draft-js                  | ^0.10.45    | ^0.11.18    |
| ◯ draft-js                         | ^0.10.5     | ^0.11.7     |
| ◯ eslint-plugin-storybook          | ^0.6.15     | ^0.11.0     |
| ◯ telepath-unpack                  | ^0.0.3      | ^0.0.4      |


### After

* 24 outdated packages (as of November 2024)

| Package                            | Old Version | New Version |
| ---------------------------------- | ----------- | ----------- |
| **Major**                          |             |             |
| ◯ @storybook/addon-controls        | ^7.6.14     | ^8.4.2      |
| ◯ @storybook/addon-docs            | ^7.6.14     | ^8.4.2      |
| ◯ @storybook/react                 | ^7.6.14     | ^8.4.2      |
| ◯ @storybook/react-webpack5        | ^7.6.14     | ^8.4.2      |
| ◯ @types/react                     | ^16.14.21   | ^18.3.12    |
| ◯ @types/react-dom                 | ^16.0       | ^18.3       |
| ◯ @typescript-eslint/eslint-plugin | ^6.21.0     | ^8.13.0     |
| ◯ @typescript-eslint/parser        | ^6.21.0     | ^8.13.0     |
| ◯ css-loader                       | ^6.10.0     | ^7.1.2      |
| ◯ cssnano                          | ^5.1.15     | ^7.0.6      |
| ◯ eslint                           | ^8.56.0     | ^9.14.0     |
| ◯ react                            | ^16.14.0    | ^18.3.1     |
| ◯ react-dom                        | ^16.14.0    | ^18.3.1     |
| ◯ react-redux                      | ^8.1.3      | ^9.1.2      |
| ◯ react-test-renderer              | ^16.14.0    | ^18.3.1     |
| ◯ react-transition-group           | ^1.1.3      | ^4.4.5      |
| ◯ sass-loader                      | ^14.1.0     | ^16.0.3     |
| ◯ storybook                        | ^7.6.14     | ^8.4.2      |
| ◯ storybook-django                 | ^0.5.1      | ^1.0.0      |
| ◯ stylelint                        | ^15.11.0    | ^16.10.0    |
| **Major version zero**             |             |             |
| ◯ @types/draft-js                  | ^0.10.45    | ^0.11.18    |
| ◯ draft-js                         | ^0.10.5     | ^0.11.7     |
| ◯ eslint-plugin-storybook          | ^0.6.15     | ^0.11.0     |
| ◯ telepath-unpack                  | ^0.0.3      | ^0.0.4      |

</details>

zerolab commented:
@lb- I am not sure upgrading draft.js to 0.11 is a good idea given https://github.com/springload/draftail/issues/213#issuecomment-826158209

lb- commented:
Good call @zerolab  - I will update my description. I did try to update but got a bunch of errors but I had not seen that issue.

lb- commented:
Description updated to be a bit clearer as to what's changed and what is just a reference of 'not yet updated'.

lb- commented:
OK, looks like we will need this PR merged to update Storybook https://github.com/torchbox/storybook-django/pull/15

If there's no traction on that soon (ping @thibaudcolas ), I will redo this PR without the Storybook update.

lb- commented:
OK, rebased and removed Storybook updates. Instead I have raised https://github.com/wagtail/wagtail/issues/12553 for that to come later when the downline dependency is updated.

lb- commented:
Draftail has now got a v2.0 update that may resolve some issues but for now I will keep this PR focused on a small set of dependencies.

Ready for review again after a fresh check of versions & a rebase.

thibaudcolas reviewed (approved):
All working as expected! Thank you @lb-

lb- commented:
Thanks @thibaudcolas
